File: apps/android/domain/src/main/kotlin/com/edge/marker/domain/DayKey.kt
package com.edge.marker.domain

import java.util.Calendar
import java.util.Date
import java.util.GregorianCalendar
import java.util.TimeZone

data class DayKey(val rawValue: String) : Comparable<DayKey> {
    constructor(year: Int, month: Int, day: Int) : this(format(year = year, month = month, day = day))

    constructor(
        date: Date,
        timeZone: TimeZone,
        calendar: Calendar = GregorianCalendar(),
    ) : this(
        year = calendar.copyWith(timeZone = timeZone).run {
            time = date
            get(Calendar.YEAR)
        },
        month = calendar.copyWith(timeZone = timeZone).run {
            time = date
            get(Calendar.MONTH) + 1
        },
        day = calendar.copyWith(timeZone = timeZone).run {
            time = date
            get(Calendar.DAY_OF_MONTH)
        },
    )

    init {
        val parts: List<String> = rawValue.split("-")
        require(parts.size == 3) { "dayKey must be formatted as YYYY-MM-DD" }

        val year: Int = parts[0].toIntOrNull() ?: error("dayKey year is invalid")
        val month: Int = parts[1].toIntOrNull() ?: error("dayKey month is invalid")
        val day: Int = parts[2].toIntOrNull() ?: error("dayKey day is invalid")

        val normalized: String = format(year = year, month = month, day = day)
        require(rawValue == normalized) { "dayKey must use canonical zero-padded format" }

        val utcCalendar: Calendar = utcCalendar()
        utcCalendar.isLenient = false
        utcCalendar.set(year, month - 1, day, 0, 0, 0)
        utcCalendar.set(Calendar.MILLISECOND, 0)
        utcCalendar.time
    }

    val date: Date
        get() = utcCalendar().run {
            val (year, month, day) = parseComponents(rawValue = rawValue)
            set(year, month - 1, day, 0, 0, 0)
            set(Calendar.MILLISECOND, 0)
            time
        }

    val weekday: MarkerWeekday
        get() = MarkerWeekday.fromCalendarValue(
            value = utcCalendar().run {
                time = date
                get(Calendar.DAY_OF_WEEK)
            },
        )

    fun addingDays(days: Int): DayKey {
        val utcCalendar: Calendar = utcCalendar()
        utcCalendar.time = date
        utcCalendar.add(Calendar.DAY_OF_MONTH, days)

        return DayKey(
            date = utcCalendar.time,
            timeZone = UTC_TIME_ZONE,
            calendar = utcCalendar,
        )
    }

    override fun compareTo(other: DayKey): Int = rawValue.compareTo(other.rawValue)

    override fun toString(): String = rawValue

    companion object {
        private val UTC_TIME_ZONE: TimeZone = TimeZone.getTimeZone("UTC")

        private fun format(year: Int, month: Int, day: Int): String = "%04d-%02d-%02d".format(year, month, day)

        private fun parseComponents(rawValue: String): Triple<Int, Int, Int> {
            val parts: List<String> = rawValue.split("-")

            return Triple(
                first = parts[0].toInt(),
                second = parts[1].toInt(),
                third = parts[2].toInt(),
            )
        }

        private fun utcCalendar(): Calendar = GregorianCalendar(UTC_TIME_ZONE)
    }
}

private fun Calendar.copyWith(timeZone: TimeZone): Calendar {
    val copy: Calendar = (clone() as Calendar)
    copy.timeZone = timeZone
    copy.isLenient = false
    return copy
}

File: apps/ios/Packages/MarkerDomain/Sources/MarkerDomain/DayKey.swift
import Foundation

public struct DayKey: RawRepresentable, Hashable, Sendable, Codable, CustomStringConvertible, Comparable {
    public let rawValue: String

    public init?(rawValue: String) {
        let parts = rawValue.split(separator: "-")
        guard parts.count == 3,
              let year = Int(parts[0]),
              let month = Int(parts[1]),
              let day = Int(parts[2]) else {
            return nil
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = DayKey.utcTimeZone

        var components = DateComponents()
        components.calendar = calendar
        components.timeZone = DayKey.utcTimeZone
        components.year = year
        components.month = month
        components.day = day

        guard calendar.date(from: components) != nil else {
            return nil
        }

        self.rawValue = DayKey.format(year: year, month: month, day: day)
    }

    public init(year: Int, month: Int, day: Int) {
        self.rawValue = DayKey.format(year: year, month: month, day: day)
    }

    public init(
        date: Date,
        timeZone: TimeZone,
        calendar: Calendar = Calendar(identifier: .gregorian)
    ) {
        var scopedCalendar = calendar
        scopedCalendar.timeZone = timeZone

        let components = scopedCalendar.dateComponents([.year, .month, .day], from: date)

        self.init(
            year: components.year ?? 1970,
            month: components.month ?? 1,
            day: components.day ?? 1
        )
    }

    public var description: String {
        rawValue
    }

    public var components: DateComponents {
        let parts = rawValue.split(separator: "-")

        return DateComponents(
            year: Int(parts[0]),
            month: Int(parts[1]),
            day: Int(parts[2])
        )
    }

    public var date: Date? {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = DayKey.utcTimeZone

        return calendar.date(from: components)
    }

    public var weekday: MarkerWeekday? {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = DayKey.utcTimeZone

        guard let date = calendar.date(from: components) else {
            return nil
        }

        return MarkerWeekday(foundationWeekday: calendar.component(.weekday, from: date))
    }

    public func addingDays(_ days: Int) -> DayKey? {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = DayKey.utcTimeZone

        guard let date,
              let shifted = calendar.date(byAdding: .day, value: days, to: date) else {
            return nil
        }

        return DayKey(date: shifted, timeZone: DayKey.utcTimeZone, calendar: calendar)
    }

    public static func < (lhs: DayKey, rhs: DayKey) -> Bool {
        lhs.rawValue < rhs.rawValue
    }

    private static let utcTimeZone = TimeZone(secondsFromGMT: 0) ?? .gmt

    private static func format(year: Int, month: Int, day: Int) -> String {
        String(format: "%04d-%02d-%02d", year, month, day)
    }
}
